### DISCUSSION TITLE
Feature Request: Frame-Level Handling for Cooldown and ActionButtonOverlay Frames

### DISCUSSION THREAD

**Discussion by Nukme** [SFX-WoW/Masque#198] [Issues] [answered]
Version: 8.2.5-beta-2

Addons: 
Dominos + Masque

Descriptions:
I keep experiencing the "cooldown covering glow" issue with _Masque_8.2.5-beta/beta2. I tried to toy with my addons and haven't found a definitive way to reproduce the problem. Sometimes it happens just after log-in, and a /reload would set things straight without disabling anything.

I use the /fstack command to track the frame levels and record them as follows:

8.0.1 normal:

<img src="https://i.ibb.co/RDJDwSv/8-0-1-normal.png" alt="8-0-1-normal" border="0">

8.2.5-beta2 normal:

<img src="https://i.ibb.co/HtJcmK9/8-2-5-normal.png" alt="8-2-5-normal" border="0">

8.2.5-beta2 abnormal:

<img src="https://i.ibb.co/88Ch9GP/8-2-5-abnormal.png" alt="8-2-5-abnormal" border="0">

As you can see, when it's normal, the _ActionButtonOverlay_ frame is always on top of the _ActionButtonCooldown_ frame.  And if _ActionButtonCooldown_ frame is higher, it would cause the GCD cooldown animation to cover glow effect.

It comes into my notice that _Masque_ 8.0.1 changes the frame level of  _ActionButtonCooldown_ frame and make it '2', while _ActionButtonOverlay_ frame is '4'. And _Masque_ 8.2.5b does not touch the frames and they are both at level '4'. My guess is that if these two frames are of the same level, the actual generation order are random. Or something else could affect the order, which I haven't figured out. 

As you mentioned in #67, _Masque_ no longer changes strata of any frames and should not be the cause of this issue, which I think it's undoubtedly right. However, by handling the frame levels as in _Masque_ 8.0.1, it could prevent this kind of issue from happening.

**Comment by Nukme** [comment]
UPDATE:

I am now affirmative that it's the frame level identity between _ActionButtonCooldown_ and  _ActionButtonOverlay_ that causes the issue.

I type in`/run ActionButton4Cooldown:SetFrameLevel(3)`  command to force the cooldown frame to be under overlay frame and the issue is gone for good. After a /reload the issue is back again. 

Strangely, if I use `/run ActionButton4Cooldown:SetFrameLevel(2)`, the cooldown frame would be under icon texture and can't be seen, which is a different behaviour from _Masque_ 8.0.1.

I've tried to read the code of _Masque_ and Blizzard's  _ActionButton.lua_ and come to a bottleneck. That is I don't know how to change the frame level of _ActionButtonOverlay_ and get it higher. It seems that _ActionButtonOverlay_ is triggered by native event and couldn't be pre-defined. 

I've tried to add a line 'Overlay:SetFrameLevel(7)' in _Core/Regions/SpellAlert.lua_ but that just cause all glow overlays on all button frames to be the same, including WAs.

I could, however,  change the frame level of cooldown frame like _Masque_ 8.0.1. But the base frame level of a Action Button is somehow protected and stays at 3. So that makes _ActionButton_ always at 3, and _ActionButtonOverlay_ always at 4. There is no room left for _ActionButtonCooldown_ .

That's all for now. I really want to solve this issue and keep _Masque_ up to date. Any thought would be helpful. Thx!

**Comment by StormFX** [comment, marked as answer]
After a bit of testing, the cooldown covering the spell alert is the _correct_ behavior. From an objective perspective, if the button is on cooldown you can't use it and the entire button, including its spell alert, should reflect that. Dominos inherits this behavior.

The reason that frame level adjustments were removed were due to their increased potential to cause taint. I can't justify reintroducing that risk for the sake of individual preference.

If you've a bit of knowledge of Lua, you can code a simple, local add-on that either lowers the cooldown frame of the cooldowns or hooks into the creation/update function for spell alerts and increase their frame level. As an alternative, Bartender4 uses a custom spell alert library and seems to do what you're asking.

**Comment by Nukme** [comment]
Thx. I'll look into that.
